[클린코드 리액트 2기 김민혁] 계산기 구현

minhyeok0328 opened the pull request:
안녕하세요 리뷰어님!
클린코드 2기 김민혁 이라고 합니다

리액트 기본 기능을 활용하여 만들었습니다.


## 기능 요구사항

- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
- [x] 2개의 숫자에 대해 곱셈이 가능하다.
- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.
- [x] 연산의 결과값이 Infinity일 경우 오류라는 문자열을 보여준다. (아이폰 참고)

## 궁금한 점

### Context API

Context API라는 것이 Provider로 value(함수, 변수 등) 값을 주입시켜서 하위 리액트 컴포넌트 들에게 전역으로 해당 값을 공유하는 것으로 이해했는데 제가 적재적소에 잘 사용한게 맞는지 궁금합니다!

### 클래스 사용

- core: 자주 사용하는 핵심적인 코드들
- domain: 앱과는 별도로 돌아가도 괜찮은? 계산만 하는 부분

위 두 폴더에 Validation이나 계산하는 부분을 단순히 값을 다루는 부분인 것 같아 반응성이랑 관련이 없어보여 클래스로 따로 분리해서 작성해 보았는데 더 좋은 방안이 있을 지 궁금합니다!

minhyeok0328 committed:
feat: 리액트 템플릿 추가 및 eslint 설정

minhyeok0328 committed:
feat: redux 추가

minhyeok0328 committed:
fix: eslint jsx 파싱에러 뜨는 문제 수정

minhyeok0328 committed:
feat: 계산기 숫자 클릭할 때 state값 업데이트 되도록 수정

minhyeok0328 committed:
fix: 0을 계속 누르면 0만 3자리 까지 추가되는 문제 수정

minhyeok0328 committed:
refactor: ts로 변경

minhyeok0328 committed:
feat: AC(All Clear)버튼을 누르면 0으로 초기화 되게 끔 수정

minhyeok0328 committed:
refactor: Main.tsx에 있는 Operator랑 AC 버튼 컴포넌트 분리

minhyeok0328 committed:
feat: 연산자 추가하는 기능 작업 및 최대 2개의 숫자에 대해 계산 가능하도록 예외처리

minhyeok0328 committed:
fix: 숫자 추가할 때 연산자 추가 이후 값 입력 안되는 오류 수정

minhyeok0328 committed:
feat: '=' 버튼 누르면 선택한 연산자에 따라 화면에 계산된 값 업데이트

minhyeok0328 committed:
feat: 계산 결과값이 Infinity 인 경우 '오류'라는 문자열로 예외처리

minhyeok0328 committed:
fix: store 변수에 TS4023 오류나는 부분 수정

minhyeok0328 committed:
refactor: 필요없는 소스코드 삭제 및 Calculator 클래스에 계산하는 함수들 private으로 선언

minhyeok0328 committed:
fix: 계산기 중앙으로 오도록 스타일 수정

minhyeok0328 committed:
refactor: redux > context api 및 useReducer로 변경

minhyeok0328 committed:
refactor: Context 파일 안에 있는 Provider 함수 따로 분리

minhyeok0328 committed:
refactor: 렌더링 최적화

minhyeok0328 committed:
feat: package.json 에 lint scripts 추가

JunilHwang commented on package.json:
```diff
@@ -0,0 +1,29 @@
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "vite-plugin-eslint": "^1.8.1"
```
`vite-plugin-eslint`도 devDependencies에 있어야 하지 않을까요!?
그렇다면 dependencies와 devDependencies의 차이점은 무엇일까요?

JunilHwang commented on vite.config.ts:
```diff
@@ -0,0 +1,14 @@
+
+// https://vitejs.dev/config/
+export default defineConfig({
+	plugins: [react(), eslint()],
```
써보질 않아서 모르겠으나... 런타임 이전에 (실행 이전에), 코드 레벨에서의 문제만 찾아주는거라면 이렇게 plugins으로 꽂아서 써야하는지 의문이 들어요!

JunilHwang replied:
조금 더 찾아보니 vite로 실행할 때 eslint를 실행해주는 역할이군요 ㅋㅋ

JunilHwang commented on src/App.tsx:
```diff
@@ -0,0 +1,21 @@
+	return (
+		<div className="App">
+			<div className="calculator">
+				<h1 id="total">{state}</h1>
```
state는 너무 범용적인 이름이라서요! `계산결과(calculation result)` 같은 이름으로 써주면 어떨까요?

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,20 @@
+		resetCalculator,
+		state
+	};
+}
```
```tsx
export default function useCalculator() {
	return useContext(CalculatorContext);
}
```
이렇게 사용해도 무방할 것 같아요!

JunilHwang commented on src/store/CalculatorContext.ts:
```diff
@@ -0,0 +1,26 @@
+		// 계산기 초기값으로 재설정
+	},
+	state: '0'
+});
```
Context에서 이렇게 함수를 별도로 관리하고 있다면,
굳이 내부적으로 useReducer를 사용할 필요가 있을까요?
단순하게 useState만 사용해도 충분할 것 같아요~

JunilHwang commented on src/store/CalculatorProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+			return;
+		}
+
+		dispatch({ type: 'insertDigits', payload: digits });
```
앞선 리뷰처럼,
지금은 불필요하게 reducer를 선언하고, 이에대한 dispatch를 사용 중인 것 같습니다.
단순하게
```tsx
const [calcResult, setCalcResult] = useState("0"")
```
이렇게 상태를 선언해주고,
이렇게 dispatch 대신에
```jsx
setCalcResult(digits);
```
이런식으로 표현해줘도 되지 않을까 싶네요!

JunilHwang commented on src/store/CalculatorProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+		}
+
+		dispatch({ type: 'insertDigits', payload: digits });
+	}, [calculatorState]);
```
여기서 생각해볼 수 있는 부분은
insertDigits가 3가지 역할을 수행중이라는 것입니다.

1. 도메인 로직 수행
2. 브라우저에 의존적인 로직 수행 (alert)
3. 상태 변경

여기서 도메인 로직과 상태 변경은 강하게 엮여있지만,
alert의 경우 굳이 provider 내에서 관리할 필요가 있을까요?

insertDigis를 수행하는 쪽에서 관리하면 되지 않을까 싶어요!

JunilHwang commented on src/store/CalculatorProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+
+export default function CalculatorProvider({ children }: { children: ReactNode }) {
+	const [state, dispatch] = useReducer(calculatorReducer, initialState);
+	const calculatorState = useMemo(() => state, [state]);
```
```jsx
const [calculatorState, dispatch] = useReducer(calculatorReducer, initialState);
```
이렇게 표현하는거랑 똑같지 않을까요?

JunilHwang commented on src/store/CalculatorProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+	}, [calculatorState]);
+
+	const insertOperation = useCallback((operator: string) => {
+		if (calculatorState === '0') {
```
`'0'` 이라는 문자열은 상수로 관리할 수 있을 것 같아요!

JunilHwang commented on src/store/calculatorReducer.ts:
```diff
@@ -0,0 +1,43 @@
+			return initialState;
+		default:
+			throw new Error(MESSAGE.NOT_FOUND_ACTION);
+	}
```
reducer를 사용할 때, case에 오는 문자열 ( type )을 상수로 만들어서 관리하면 편하답니다! 무엇보다 휴면 에러를 방지할 수 있어요 ㅎㅎ

JunilHwang commented on src/store/CalculatorProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+		dispatch({ type: 'setAnswer', payload: item.result() });
+	}, [calculatorState]);
+
+	const resetCalculator = () => dispatch({ type: 'resetCalculator' });
```
이만큼의 코드도 별도의 hook으로 분리해서 관리하면 더 좋을 것 같아요!
지금은 Provider 하나가 너무 비대해진 상태라서 코드를 이해하기가 어려울 수 있답니다 ㅠㅠ

JunilHwang commented on src/store/CalculatorProvider.tsx:
```diff
@@ -0,0 +1,66 @@
+		>
+			{children}
+		</CalculatorContext.Provider>
+	);
```
리액트의 컴포넌트는 "얕은 비교"를 수행하면서 "리렌더링"을 판단합니다.

"얕은 비교"는 값이 동일한 "메모리"에 위치하고 있는지 보는건데요,

지금처럼 value에 각각의 property를 직접 할당할 경우 무조건 다른 메모리에 값을 할당하게 됩니다.

이를 해결하기 위해서 어떻게 하면 좋을까요?

JunilHwang commented on src/core/Validator.ts:
```diff
@@ -0,0 +1,18 @@
+
+		return text.length <= MAX_DIGIT_LENGTH;
+	}
+}
```
잘 생각해보면, 이것도 core가 아니라 domain에 해당하는 로직이 아닐까요?

JunilHwang commented on src/domain/Calculator.ts:
```diff
@@ -0,0 +1,29 @@
+	constructor(operator: string, calculateData: number[]) {
+		this.operator = operator;
+		this.calculateData = calculateData;
+	}
```
```typescript
  constructor(
    private readonly operator: string,
    private readonly calculateData: number[],
  ) {}
```
이렇게 표현할 수 있답니다!

JunilHwang commented on src/domain/Calculator.ts:
```diff
@@ -0,0 +1,29 @@
+			default:
+				throw new Error(`'${this.operator}' 연산자는 정의되지 않은 연산자입니다.`);
+		}
+	}
```
여기서 사용되는 연산자들의 경우에도 상수로 관리할 수 있지 않을까요?

JunilHwang replied:
```typescript
result() {
  Calculator.operators.get(this.operator)?.result();
}
```
앞선 리뷰를 토대로 표현해보자면, 이런 느낌임입니다!

JunilHwang commented on src/domain/Calculator.ts:
```diff
@@ -0,0 +1,29 @@
+	private plus = (): number => this.calculateData.reduce((acr, cur) => acr + cur);
+	private minus = (): number => this.calculateData.reduce((acr, cur) => acr - cur);
+	private multiply = (): number => this.calculateData.reduce((acr, cur) => acr * cur);
+	private divide = (): number => Math.round(this.calculateData.reduce((acr, cur) => acr / cur));
```
"연산자(Operator)" 라는 클래스가 있고,
연산자 기호를 가지고
숫자들을 입력 받으면 연산을 수행하는 메소드가 있다면,

result와 이 부분을 더 효과적으로 표현할 수 있을 것 같아요!

JunilHwang commented on src/App.tsx:
```diff
@@ -0,0 +1,21 @@
+				<h1 id="total">{state}</h1>
+				<DigitsContainer/>
+				<ResetContainer/>
+				<OperationContainer/>
```
이 컴포넌트들은, 사실 "컴포넌트"의 역할을 제대로 수행하고 있지 못한게 아닐까요?
왜냐하면 main.tsx에서 provider를 씌워줬고,
결국 이 컨테이너들이 provider를 가져다 사용하고 있습니다.
provider에 의존적인 상황인거죠.

그래서 이름을 container로 지은 것 같은데,
그렇다면 containers 라는 폴더를 만든 후에 containers역할을 수행하는 컴포넌트를 모아놓아야 한다고 생각합니다!

JunilHwang commented on src/App.tsx:
```diff
@@ -0,0 +1,21 @@
+import DigitsContainer from '@/components/DigitsContainer';
+import ResetContainer from '@/components/ResetContainer';
+import OperationContainer from '@/components/OperationContainer';
+import useCalculator from '@/hooks/useCalculator';
```
```tsx
import { DigitsContainer, ResetContainer, OperationContainer } from '@/containers';
import { useCalculator } from '@/hooks';
```
이런식으로 사용하기 위해선 폴더 구조를 어떻게 정의해줘야 할까요!?

JunilHwang commented on src/components/Digits.tsx:
```diff
@@ -0,0 +1,17 @@
+			onClick={() => insertDigits(digits)}
+		>{digits}</button>
+	);
+}
```
```typescript
export default function Digits({ digits, onClick }: Props) {
    const handleClick = useCallback(() => { onClick(digis) }, [onClick])

	return (
		<button
			className="digit"
			type="button"
			onClick={handleClick}
		>{digits}</button>
	);
}
```
이렇게 컴포넌트를 정의해준다면 provider/context 에 독립적인 형태로 사용할 수 있을 것 같아요!

무엇보다 앞서 언급했던 "provider"가 모두 렌더링 된다는건, 컴포넌트에도 영향이 있다는 이야기라서요.

그래서 context와 provider를 사용할 때, 어떻게 하면 렌더링을 최적화할 수 있을지 고민이 꼭 필요하답니다!

JunilHwang commented on src/components/DigitsContainer.tsx:
```diff
@@ -0,0 +1,10 @@
+			{DIGIT.map((digits, index) => (<Digits key={index} digits={digits}/>))}
+		</div>
+	);
+}
```
이 컴포넌트가 Container이기 때문에,
여기서 context에 있는 것들을 가져와서 Digits에게 넘겨줘야할 것 같아요!

JunilHwang commented on src/components/DigitsContainer.tsx:
```diff
@@ -0,0 +1,10 @@
+export default function DigitsContainer() {
+	return (
+		<div className="digits flex">
+			{DIGIT.map((digits, index) => (<Digits key={index} digits={digits}/>))}
```
`DIGIT`라는 상수는 여기에서만 쓰이는 것 같아요.
그렇다면 이 파일 내에 이 상수가 선언되어야 이해하기도, 관리하기도 쉬울 것 같습니다!

JunilHwang commented on src/components/Operation.tsx:
```diff
@@ -0,0 +1,25 @@
+	const { setAnswer, insertOperation } = useCalculator();
+
+	const clickOperatorHandler = () => {
+		if (operator === '=') {
```
`=` 도 상수로 관리할 수 있을 것 같네요 ㅎㅎ

JunilHwang commented on src/components/Operation.tsx:
```diff
@@ -0,0 +1,25 @@
+			onClick={clickOperatorHandler}
+		>{operator}</button>
+	);
+}
```
이 컴포넌트도 context에 독립적인 형태로 만들 수 있을 것 같습니다!

JunilHwang commented on src/components/OperationContainer.tsx:
```diff
@@ -0,0 +1,15 @@
+			<Operation operator="="/>
+		</div>
+	);
+}
```
앞선 리부를 참고해주세요!

JunilHwang commented on src/constants/index.ts:
```diff
@@ -0,0 +1,10 @@
+	NOT_FOUND_ACTION: 'action 을 찾을 수 없습니다.',
+};
+export const DIGIT = ['9', '8', '7', '6', '5', '4', '3', '2', '1', '0'];
+export const OPERATOR = ['/', 'X', '-', '+'];
```
배열이기 때문에 DIGITS, OPERATORS 처럼 표현해줘야 맞을 것 같아요!

JunilHwang commented on src/constants/index.ts:
```diff
@@ -0,0 +1,10 @@
+	MAX_OPERATOR_LENGTH: `최대 ${MAX_OPERATOR_LENGTH + 1}개의 숫자에 대해서만 계산할 수 있습니다.`,
+	INFINITY: '오류',
+	NOT_FOUND_ACTION: 'action 을 찾을 수 없습니다.',
+};
```
DIGIT_LENGTH와 OPERATOR_LENGTH의 경우 "도메인"을 표현하는 상수고,
메세지는 UI를 표현하는 상수입니다.
서로 역할이 다르기 때문에 다른 장소에서 다른 방식으로 관리되어야 하지 않을까요!?

무엇보다 MAX_DIGIT_LENGTH와 MAX_OPERATOR_LENGTH는 일종의 config로서 관리될 수 있을 것 같아요.
즉, 외부에서 주입해줄 수 있는 형태로 관리해주면 더 좋지 않을까 싶습니다 ㅎㅎ

JunilHwang commented on src/components/ResetContainer.tsx:
```diff
@@ -0,0 +1,11 @@
+			<button className="modifier" onClick={resetCalculator}>AC</button>
+		</div>
+	);
+}
```
`resetCalculator` 라는 함수를 props로 받아오면, 이 컴포넌트도 context에 독립적인 형태로 사용할 수 있을 것 같네요!

JunilHwang commented on src/App.tsx:
```diff
@@ -0,0 +1,21 @@
+			</div>
+		</div>
+	);
+}
```
앞선 리뷰들을 참고해봤을 때, 지금은 Provider와 Context가 거의 불필요한 상황입니다.
사실 학습 목적으로 사용한게 있는 것 같아요!
이제 사용해봤으니, 사용하지 않아야 하는 경우도 인지했기를 바랍니다 🙏

JunilHwang reviewed (changes requested):
안녕하세요 민혁님!
계산기 미션을 함께하게 된 리뷰어 황준일입니다 🙏
미션 무척 빨리 수행해주셨네요 😁

전체적으로 확인해볼 부분은
컴포넌트는 최대한 "독립적"인 형태로 쓰일 수 있도록 만들어야 한다고 생각합니다.
그래야 "컴포넌트"답게 사용할 수 있으니까요!
지금은 context에서 함수와 값을 꺼내오지 않고, 단순하게 props로 건내받아서 사용해도 무방한 구조 같아요!
"외부 환경에 독립적인 컴포넌트"로 만들어서 관리해도 충분할 것 같습니다.
지금은 모든 컴포넌트가 context에 의존적인 상태이기 때문에 만약에 요구사항이 변경되거나 했을 때 대응하기가 무척 어려울 것 같아요!

이 과정은 "클린코드"를 추구하는 과정이기 때문에 "클린코드"라는게 뭔지 한 번 고민해보시고, 고민한 내용을 토대로 코드를 작성해가면 될 것 같습니다 ㅎㅎ

자세한건 코드리뷰 내용을 확인해주세요!

그리고 질문에 대한 답변을 드리자면,

> Context API라는 것이 Provider로 value(함수, 변수 등) 값을 주입시켜서 하위 리액트 컴포넌트 들에게 전역으로 해당 값을 공유하는 것으로 이해했는데 제가 적재적소에 잘 사용한게 맞는지 궁금합니다!

"전역" 이라기보단, Context를 공유한다고 보시면 좋을 것 같아요! App에다 감싸면 전역이지만,
필요한 영역에다만 감싸면 "전역"이 아니기 때문입니다.

지금은 reducer와 같이 사용해주셨는데, 지금 꼭 reducer가 필요한가? 에 대해서도 고민해보시면 좋을 것 같습니다. 지금은 useState만 사용해도 충분해보여요!

> 클래스 사용 관련 (core, domain)

class로 분리한건 무척 좋아보입니다! 다만 core에 있는 validator도 domain의 한 축인 것 같아요. core라는 이름에 어울리지 않는다고 생각합니다. 그렇다면 어떤 것들이 domain이 될 수 있는지 고민해보시면 좋을 것 같아요!

minhyeok0328 committed:
refactor: 기존 Digits, Operation 컴포넌트 리펙터링

순수함수로써 동작하도록 Container 컴포넌트메서 state와 함수 props로 전달하도록 수정

minhyeok0328 committed:
refactor: container 컴포넌트는 container 폴더로 이동

추가로 compoents 폴더랑 container폴더에 index.ts 추가해서 export default as로 return하개 해서 가져오기 편하게 수정

minhyeok0328 committed:
refactor: useReduce > useState로 변경

기존에 Calculate 관련 함수들은 store/hooks에 따로 분리

minhyeok0328 committed:
refactor: Context API 사용한 부분들 제거

minhyeok0328 committed:
fix: state 값이 변경될 때 re-rendering 되지 않는 문제 수정

Container에서 useCalculator 훅을 호출하게 되면 각각 다른 state가 생성되므로
App.tsx에서 함수랑 state 값을 props로 넘겨주도록 수정

minhyeok0328 committed:
refactor: domain 로직 분리

1. Validator.ts 파일 의존성 제거 후 domain 폴더로 이동
2. 기존 Calculator.ts 파일에서 쓰던 계산로직들 Operator.ts 파일 새로 생성 후에 해당 파일에 분리
3. 기존 constants/index.ts에 있는

MAX_DIGIT_LENGTH
MAX_OPERATOR_LENGTH

와 같은 상수들을 각각 맞는 도메인 클래스로 이동

minhyeok0328 committed:
fix: 계산기 operator 순서 반대로 바뀐부분 수정

minhyeok0328 committed:
refactor: 코드 라인 한 줄로 긴 부분들 개행처리 및 상수 불러오는 부분들 정리

minhyeok0328 committed:
fix: vite-plugin-eslint을 devDependencies로 이동

minhyeok0328 committed:
refactor: Digits > Digit로 컴포넌트 이름 변경
나머지는 간단한 코드 정리

minhyeok0328 committed:
refactor: Operator.ts에 SYMBOLS 부분 Map으로 변경

minhyeok0328 committed:
refactor: digits > digit 으로 변수 및 타입 이름 변경

minhyeok0328 committed:
refactor: useCalculator에 있는 alert 실행 위치 변경

Validator가 통과하지 못했을 때 alert를 띄워주는걸
훅 함수를 실행하는 쪽으로 이동
> alert가 브라우저 의존적인 로직이라서

JunilHwang commented on src/components/Digit.tsx:
```diff
@@ -0,0 +1,18 @@
+}
+
+export default function Digit({ digit, onClick }: Props) {
+	const handlerOnClick = () => {
```
`handleClick` 이라고 표현해야 자연스러울 것 같아요!

JunilHwang commented on src/components/Operation.tsx:
```diff
@@ -0,0 +1,17 @@
+}
+
+export default function Operation({ operator, onClick }: Props) {
+	const handlerOnClick = () => {
```
여기도 동일합니다~

JunilHwang commented on src/components/Operation.tsx:
```diff
@@ -0,0 +1,17 @@
+	onClick: (operation: string) => void;
+}
+
+export default function Operation({ operator, onClick }: Props) {
+	const handlerOnClick = () => {
+		onClick(operator);
+	};
+
+	return (
+		<button
+			className="operation"
+			onClick={handlerOnClick}
+		>{operator}</button>
+	);
+}
```
지금 Digit이랑 Operation이 똑같은 구조로 되어있어요!
이를 추상화할 수 있지 않을까요?
```tsx
<CalculatorButton type="digit" onClick="handleClickDigit">1</CalculatorButton>
<CalculatorButton type="operation" onClick="handleClickOperation">+</CalculatorButton>
```
이런 느낌입니다!

JunilHwang commented on src/constants/index.ts:
```diff
@@ -0,0 +1,5 @@
+export const MESSAGE = {
+	MAX_DIGIT_LENGTH: (length: number) => `숫자는 한번에 최대 ${length}자리 수까지 입력 가능합니다.`,
+	MAX_OPERATOR_LENGTH: (length: number) => `최대 ${length}개의 숫자에 대해서만 계산할 수 있습니다.`,
```
이건 상수라기 보단 함수라서
그냥 이 Message가 쓰이는 컴포넌트 내에서 표현해주는게 좋지 않을까 싶습니다!

JunilHwang commented on src/container/DigitsContainer.tsx:
```diff
@@ -0,0 +1,34 @@
+export default function DigitsContainer({ insertDigit, calculatorState }: Props) {
+	const handleInsertDigit = (digit: string) => {
+		if (!Validator.isOverMaxDigitLength(calculatorState + digit)) {
+			alert(MESSAGE.MAX_DIGIT_LENGTH(MAX_DIGIT_LENGTH));
```
`MAX_DIGIT_LENGTH(MAX_DIGIT_LENGTH)`
뭔가 이상하지 않나요!? 함수 이름과 변수 이름이 똑같아요.

그래서 
```typescript
invalidMaxDigitLength(MAX_DIGIT_LENGTH);
```
이런식으로 표현해주면 어떨까요?

JunilHwang commented on src/container/DigitsContainer.tsx:
```diff
@@ -0,0 +1,34 @@
+
+
+export default function DigitsContainer({ insertDigit, calculatorState }: Props) {
+	const handleInsertDigit = (digit: string) => {
+		if (!Validator.isOverMaxDigitLength(calculatorState + digit)) {
+			alert(MESSAGE.MAX_DIGIT_LENGTH(MAX_DIGIT_LENGTH));
+			return;
+		}
+
+		insertDigit(digit);
+	};
```
함수 이름은 `handleClickDigit` 처럼 표현하면 적절할 것 같아요!

"digit을 click했을 때 실행한다" 라는 의미입니다.

JunilHwang commented on src/container/OperationContainer.tsx:
```diff
@@ -0,0 +1,37 @@
+const { symbols: OPERATOR_SYMBOLS, MAX_OPERATOR_LENGTH } = Operator;
+
+export default function OperationContainer({ setAnswer, insertOperation, calculatorState }: Props) {
+	const handleInsertOperation = (operator: string) => {
+		if (Validator.isMaxOperatorLength(calculatorState + operator)) {
+			alert(MESSAGE.MAX_OPERATOR_LENGTH(MAX_OPERATOR_LENGTH));
+			return;
+		}
+
+		insertOperation(operator);
+	};
+
+	return (
+		<div className="operations subgrid">
+			{OPERATOR_SYMBOLS.map((operator: string) => (
+				<Operation
+					key={operator}
+					onClick={handleInsertOperation}
```
앞선 내용과 동일합니다!

JunilHwang commented on src/container/OperationContainer.tsx:
```diff
@@ -0,0 +1,37 @@
+					operator={operator}/>
+			))}
+			<Operation
+				operator="="
```
`=` 이라는 기호도 상수로 관리할 수 있을 것 같아요..!

JunilHwang commented on src/container/ResetContainer.tsx:
```diff
@@ -0,0 +1,11 @@
+	resetCalculator: () => void;
+}
+
+export default function ResetContainer({ resetCalculator }: Props) {
+	return (
+		<div className="modifiers subgrid">
+			<button className="modifier" onClick={resetCalculator}>AC</button>
+		</div>
+	);
+}
```
이 컴포넌트한테는 memo를 씌우면 좋을 것 같아요! reset은 항상 변하지 않는 함수일 것 같거든요.

```typescript
function ResetContainer({ resetCalculator }: Props) {
	return (
		<div className="modifiers subgrid">
			<button className="modifier" onClick={resetCalculator}>AC</button>
		</div>
	);
}

export default memo(ResetContainer);
```

이렇게요!

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,60 @@
+export default function useCalculator() {
+	const [calculatorState, setState] = useState(initialState);
+
+	const insertDigit = useCallback((digit: string) => {
+		if (calculatorState === initialState) {
+			setState(digit);
+			return;
+		}
+
+		setState(calculatorState + digit);
+	}, [calculatorState]);
+
+	const insertOperation = useCallback((operator: string) => {
+		if (calculatorState === initialState) {
+			return;
+		}
+
+		setState(calculatorState + operator);
+	}, [calculatorState]);
```
이 두 함수가 잘 보면 똑같은 것 같지 않나요?
인자의 이름만 다른 것 같아요!

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,60 @@
+	}, [calculatorState]);
+
+	const setAnswer = useCallback(() => {
+		const operator = calculatorState.split('').find((item) => OPERATOR_SYMBOLS.includes(item));
```
```typescript
const operator = calculatorState.split('').find((item) => Operator.SYMBOLS.get(item));
```
이렇게 표현해도 되지 않을까요!?

JunilHwang replied:
앞선 리뷰처럼, `Operator.findBy(calculatorState)` 이렇게 쓰일 수 있는 메소드를 하나 만들어보면 어떨까요? Validator에서도 적절하게 사용될 수 있을 것 같아요!

JunilHwang commented on src/domain/Operator.ts:
```diff
@@ -0,0 +1,25 @@
+export default class Operator {
+	public static readonly MAX_OPERATOR_LENGTH = 1;
+	public static readonly SYMBOLS = new Map<string, Operator>([
+		['+', new Operator('+', (a, b) => a + b)],
+		['-', new Operator('-', (a, b) => a - b)],
+		['X', new Operator('X', (a, b) => a * b)],
+		['/', new Operator('/', (a, b) => Math.round(a / b))],
+	]);
```
이미 `new Operator`에 symbol을 넘겨주고 있어서, 

```typescript
public static readonly SYMBOLS = [
    new Operator('+', (a, b) => a + b),
    new Operator('-', (a, b) => a - b),
    new Operator('X', (a, b) => a * b),
    new Operator('/', (a, b) => Math.round(a / b)),
].reduce((map, value) => {
  map.add(value.operator, value);
  return map
}, new Map<string, Operator>());
```

이런식으로 표현해주면 어떨까요!?

JunilHwang commented on src/domain/Operator.ts:
```diff
@@ -0,0 +1,25 @@
+		['+', new Operator('+', (a, b) => a + b)],
+		['-', new Operator('-', (a, b) => a - b)],
+		['X', new Operator('X', (a, b) => a * b)],
+		['/', new Operator('/', (a, b) => Math.round(a / b))],
```
요구사항에서는 "반올림"이 아니라 "버림" 입니다!

JunilHwang commented on src/domain/Validator.ts:
```diff
@@ -0,0 +1,27 @@
+import { Calculator, Operator } from './index';
+
+const { symbols: OPERATOR_SYMBOLS, MAX_OPERATOR_LENGTH } = Operator;
+const { MAX_DIGIT_LENGTH } = Calculator;
```
합칠 수 있을 것 같아요!

JunilHwang commented on src/domain/Validator.ts:
```diff
@@ -0,0 +1,27 @@
+
+export default class Validator {
+	static isMaxOperatorLength(text: string): boolean {
+		const findOperator = text.split('').filter((item: string) => (
+			OPERATOR_SYMBOLS.includes(item)
+		));
```
이건 Operator의 method로 관리할 수 있지 않을까요?

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,60 @@
+			return;
+		}
+
+		const [a, b] = calculatorState.split(operator).map((item) => Number(item));
+		const item = new Calculator(
+			operator,
+			[a, b]
+		);
```
```typescript
const item = new Calculator(
    operator,
    calculatorState.split(operator).map(Number)
);
```
이렇게 표현할 수 있겠네요!

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,60 @@
+		);
+		const result = item.execute();
+
+		if (result === Infinity) {
```
-Infinity 일 때도 있지 않을까요?

`-1/0` 같은 연산을 수행할 수도 있으니까요!

JunilHwang replied:
방금 테스트 해보니까, 결과값이 음수면 추가 연산이 아예 불가능하네요.. ㅎㅎ 음수 입력도 안 되는군요..!

JunilHwang commented on src/hooks/useCalculator.ts:
```diff
@@ -0,0 +1,60 @@
+	}, [calculatorState]);
+
+	const insertOperation = useCallback((operator: string) => {
+		if (calculatorState === initialState) {
+			return;
+		}
```
연산 결과가 0일 경우에는 추가 연산이 아예 불가능해져버리네요..!

JunilHwang reviewed (approved):
안녕하세요 민혁님!
피드백 반영 잘 해주셨네요 👍
다만 아직 고려해야될 부분이 몇 가지 보이는 것 같아요.
어플리케이션의 버그도 몇 개 발견했네요 ㅎㅎ

계산기 미션은 온보딩 미션이기 때문에 여기서 더 깊게 들어가진 않도록 하겠습니다!
이번에 남긴 피드백은 다음 미션 진행하실 때 고려해보시면 좋을 것 같아요!

부디 제가 민혁님께 조금이나마 도움이 되었기를 바랍니다 🙏
끝까지 무사완주 하기를 기대하겠습니다 😉
화이팅! 💪💪
